### PR TITLE
Choose chars

### PR DESCRIPTION
**Pull request by yonbergman**
Solving   #8 
Using a gridview and better character selection
screen shot 2015-03-03 at 9 12 45 pm
screen shot 2015-03-03 at 9 12 59 pm

### CODE REVIEW

File: app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        private Integer mCharacterB;
+
+        public ChooseCharactersFragment() {
+            setHasOptionsMenu(true);
+        }
+
+        @Override
+        public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+            super.onActivityCreated(savedInstanceState);
+            //TODO: Load player from intent.
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java]: I think you can write this code, even if no player is being sent yet.

File: app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            }
+            refreshDoneMenuItem();
+        }
+
+        private void refreshDoneMenuItem(){
+            mDoneMenuItem.setEnabled(validSelectionCount());
+        }
+        
+        private boolean validSelectionCount(){
+//            Log.i("SELECTCHARS", String.format("%s %s", String.valueOf(mCharacterA), String.valueOf(mCharacterB)));
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java]: remove

File: app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            return super.onOptionsItemSelected(item);
+        }
+
+        private void performDone() {
+            Pair<Character, Character> characterPair = selectedCharacters();
+            Task<Player> setCharacterTask = new SetCharactersOperation().run(mPlayer, characterPair.first, characterPair.second);
+            setCharacterTask.continueWith(new Continuation<Player, Object>() {
+                @Override
+                public Object then(Task<Player> task) throws Exception {
+                    Player player = task.getResult();
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java:96]: You didn't check for failures

**Comment by udiWertheimer** [comment]
screenshot 2015-03-04 10 12 51

File: app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        @Override
+        public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                                 Bundle savedInstanceState) {
+            View rootView = inflater.inflate(R.layout.choose_character_fragment, container, false);
+            ButterKnife.inject(this, rootView);
+
+            mCharacterAdapter = new CharacterAdapter(getActivity());
+            mCharacterListView.setAdapter(mCharacterAdapter);
+            mCharacterListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java]: User ButterKnife's event handling capability

File: app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        public CharacterAdapter(Context context) {
+            super(context, new QueryFactory<Character>() {
+                @Override
+                public ParseQuery<Character> create() {
+                    ParseQuery<Character> query = ParseQuery.getQuery(Character.class);
+                    query.orderByAscending("name");
+                    return query;
+                }
+            });
+            setObjectsPerPage(60);
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java]: Magic number?

File: app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+        // Customize the layout by overriding getItemView
+        @Override
+        public View getItemView(Character character, View v, ViewGroup parent) {
+            if (v == null) {
+                v = View.inflate(getContext(), R.layout.character_list_item, null);
+            }
+            super.getItemView(character, v, parent);
+
+            TextView titleTextView = (TextView) v.findViewById(R.id.character_name);
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/ChooseCharactersActivity.java]: ViewHolder pattern plz

File: app/src/main/java/com/iic/mokojin/operations/SetCharactersOperation.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    public Task<Player> run(Player player, Character characterA, Character characterB){
+        HashMap<String, Object> params = buildParams(player, characterA, characterB);
+        return ParseCloud.callFunctionInBackground(SET_CHARACTER_CLOUD_FUNCTION_NAME, params);
+    }
+    
+    private HashMap<String, Object> buildParams(Player player, com.iic.mokojin.models.Character characterA, Character characterB){
+        HashMap<String, Object> params = new HashMap<>();
+        //TODO: Remove fallback when no player
+        params.put("player",  null != player ? player.getObjectId() : "PI9LhRWxxa"); //
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/operations/SetCharactersOperation.java:27]: I don't think that string needs to be there :)

File: app/src/main/java/com/iic/mokojin/presenters/CharacterPresenter.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+import android.content.Context;
+
+/**
+ * Created by yon on 3/3/15.
+ */
+public class CharacterPresenter {
+    
+    public static int getImageResource(Context context, com.iic.mokojin.models.Character character){
+        String resourceId = "player_".concat(String.valueOf(character.getCharacterId()));
```
Comment (giladgo) [inline app/src/main/java/com/iic/mokojin/presenters/CharacterPresenter.java:11]: Any reason not to use `+`?

**Comment by giladgo** [comment]
:dancer:

File: app/src/main/java/com/iic/mokojin/presenters/CharacterPresenter.java (hunk 8 above)
Comment (yonbergman) [inline app/src/main/java/com/iic/mokojin/presenters/CharacterPresenter.java:11]: concat is suppose to be faster

**Pull request merged**